Add `sync` module with a mutex implementation.

wedsonaf opened the pull request:
This introduces a `sync` module with a mutex implementation. In the next couple of days I'll send pull requests for spinlocks, condvars, and locked_by. These are all used in the driver: I'm planning to submit all rust/ changes before the driver.

ojeda commented on rust/kernel/sync/guard.rs:
```diff
@@ -0,0 +1,81 @@
+
+// SAFETY: `Guard` is sync when the data protected by the lock is also sync. This is more
+// conservative than the default compiler implementation; more details can be found on
+// https://github.com/rust-lang/rust/issues/41622 -- it refers to MutexGuard from the standard
```
```suggestion
// https://github.com/rust-lang/rust/issues/41622 -- it refers to `MutexGuard` from the standard
```

wedsonaf replied:
Done.

ojeda commented on rust/kernel/sync/mutex.rs:
```diff
@@ -0,0 +1,95 @@
+}
+
+// SAFETY: `Mutex` can be transferred across thread boundaries iff the data it protects can.
+unsafe impl<T: ?Sized + Send> Send for Mutex<T> {}
+// SAFETY: `Mutex` serialises the interior mutability it provides, so it is `Sync` as long as the
```
Empty line to separate this, please?

wedsonaf replied:
Done.

ojeda commented on rust/kernel/sync/mutex.rs:
```diff
@@ -0,0 +1,95 @@
+use crate::{bindings, CStr};
+use core::{cell::UnsafeCell, pin::Pin};
+
+/// Safely initialises a [`Mutex`] wih the given name and new lock class.
```
```suggestion
/// Safely initialises a [`Mutex`] with the given name, generating a new lock class.
```

i.e. it sounded like the new lock class is also passed rather than generated inside.

wedsonaf replied:
Done.

ojeda commented on rust/kernel/sync/mod.rs:
```diff
@@ -0,0 +1,20 @@
+    ($obj:expr, $name:literal) => {{
+        static mut CLASS: core::mem::MaybeUninit<$crate::bindings::lock_class_key> =
+            core::mem::MaybeUninit::uninit();
+        // SAFETY: `CLASS` is never used by rust code directly; the kernel may change it though.
```
```suggestion
        // SAFETY: `CLASS` is never used by Rust code directly; the kernel may change it though.
```

wedsonaf replied:
Done.

ojeda commented on rust/kernel/sync/mod.rs:
```diff
@@ -0,0 +1,20 @@
+
+/// Safely initialises an object that has an `init` function that takes a name and a lock class as
+/// arguments, examples of these are [`Mutex`], [`SpinLock`], and [`Condvar`]. Each of them also
+/// provides a more specialised name that uses this macro.
```
Should we provide this if the others will provide the more specialized macro?

wedsonaf replied:
The name is more specialised but it just calls `init_with_lockdep`. For example, one would call `mutex_init` instead of `init_with_lockdep` though both would work.

ojeda replied:
What I meant is that we could hide `init_with_lockdep!` unless users need that one. The macro needs to be exported because it is called from a user-facing macro, but we could still do `#[doc(hidden)]`.

wedsonaf replied:
Oh, I see. I've added `#[doc(hidden)]`.

ojeda commented on rust/kernel/sync/guard.rs:
```diff
@@ -0,0 +1,81 @@
+// SPDX-License-Identifier: GPL-2.0
+
```
```suggestion

//! Lock guard.

```

i.e. please document the modules themselves, even if not public -- the idea is to use it as a "file information header".

wedsonaf replied:
Done.

ojeda commented on rust/kernel/sync/mod.rs:
```diff
@@ -0,0 +1,20 @@
+// SPDX-License-Identifier: GPL-2.0
+
```
Missing doc.

wedsonaf replied:
Done.

ojeda commented on rust/kernel/sync/mutex.rs:
```diff
@@ -0,0 +1,95 @@
+    };
+}
+
+/// Exposes the kernel's `struct mutex`. When multiple threads attempt to lock the same mutex, only
```
```suggestion
/// Exposes the kernel's [`struct mutex`]. When multiple threads attempt to lock the same mutex, only
```

plus at the bottom:

```
/// [`struct mutex`]: ../../../../include/linux/mutex.h
```

I might have got the number of `..`s wrong :-)

wedsonaf replied:
Done.

ojeda reviewed (commented):
Quick usual doc review + some CI failures too.

alex commented on rust/kernel/sync/mod.rs:
```diff
@@ -0,0 +1,24 @@
+        static mut CLASS: core::mem::MaybeUninit<$crate::bindings::lock_class_key> =
+            core::mem::MaybeUninit::uninit();
+        // SAFETY: `CLASS` is never used by Rust code directly; the kernel may change it though.
+        unsafe { $obj.init($crate::cstr!($name), CLASS.as_mut_ptr()) };
```
Can we do something to assert the type of `$obj` -- maybe add an `_: &TYPE: &$obj;`  or something? Otherwise this feels a bit dangerous.

wedsonaf replied:
This is meant to work on multiple different and unrelated types, for example, `Mutex`, `SpinLock` and `Condvar`.

How is this dangerous? (It's a genuine question, I don't quite get the dangerous aspect.)

I suppose I could define a trait that needs to be implemented by structs with such a pattern, but I'd like to understand what we gain from it. At the moment, if this macro is called on an object that doesn't have `init` or where `init` has a different type, it will just fail to compile.

Do you know of better approaches?

alex replied:
So, the danger would be something like: you have a method called init on an unrelated type, which has the same signature, but different invariants. This would let you call that without `unsafe` appearing in your source code. It's a bit theoretical, I'll grant you.

I think poor error messages is probably a bigger issue in practice :-)

ojeda replied:
I thought the macro was being used to generate the `static`, rather than trying to imitate duck typing or C++ templates ;-P If a trait would do, it'd be better than a macro.

ojeda replied:
Anyhow, if the macro is kept like this, we should note the macro is "unsafe" and have the usual comment (I am doing something similar for the printing stuff I am uploading).

wedsonaf replied:
I see the danger now. Indeed we need to protect users from this.

@ojeda -- we need the macro to generate the static class, the trait would be in addition to the macro. It would be used to ensure that this only works when called on objects of the right type.

alex commented on rust/kernel/sync/mutex.rs:
```diff
@@ -0,0 +1,102 @@
+
+// SAFETY: `Mutex` serialises the interior mutability it provides, so it is `Sync` as long as the
+// data it protects is also `Sync`.
+unsafe impl<T: ?Sized + Sync> Sync for Mutex<T> {}
```
Should this be?

```suggestion
unsafe impl<T: ?Sized + Send> Sync for Mutex<T> {}
```

That's what stdlib mutex is: https://doc.rust-lang.org/stable/std/sync/struct.Mutex.html#impl-Sync

wedsonaf replied:
Nice catch, thank you! Indeed we only need `T` to be `Send`.

alex reviewed (commented):
Can you add a full example of using the mutex (particularly the init stuff) to the module comment.

ojeda commented:
> Can you add a full example of using the mutex (particularly the init stuff) to the module comment.

+1, and please copy-paste it into `rust_example.rs`, since those are effectively our tests for the moment.

ojeda commented:
Why `Mutex` isn't `PhantomPinned`? Its `init()` requires a pinned reference, but that doesn't mean much if one can pin and unpin at will, no?

Related: it might be a good idea to document why we require a type to be pinned or not (e.g. `x` member from `y` C structure is self-referential), even if it only applies under some `CONFIG` option (e.g. debug) etc.

wedsonaf commented:
> Why `Mutex` isn't `PhantomPinned`? Its `init()` requires a pinned reference, but that doesn't mean much if one can pin and unpin at will, no?

Yes, I'm adding `PhantomPinned` to all types that require pinning.
 
> Related: it might be a good idea to document why we require a type to be pinned or not (e.g. `x` member from `y` C structure is self-referential), even if it only applies under some `CONFIG` option (e.g. debug) etc.

Sure, I'll add some text around this.

wedsonaf commented:
> > Can you add a full example of using the mutex (particularly the init stuff) to the module comment.

I added a trivial example.

> +1, and please copy-paste it into `rust_example.rs`, since those are effectively our tests for the moment.

This is not very useful in `rust_example.rs` -- I'll eventually add something there, but at the moment I'm a bit busy with getting the existing code in shape for the merge window.

ojeda commented:
> This is not very useful in `rust_example.rs` -- I'll eventually add something there, but at the moment I'm a bit busy with getting the existing code in shape for the merge window.

It is useful because it is the only way we have to at least compile-test the macros and run at least the happy path once in the CI.

If you are busy, that is fine, I can add it later myself :-)

wedsonaf commented:
> It is useful because it is the only way we have to at least compile-test the macros and run at least the happy path once in the CI.

Fair enough. I've added some code there for this purpose, though it is not a useful sample of using mutexes.

(Later on I'd like to add something like a counter of the number of bytes written, then `read` will return this count. Not useful functionality but shows how a mutex could be used. I need to bring some more changes to `rust/` before I can write this.)

wedsonaf commented:
Folks, thank you for the reviews. I believe I addressed all the comments.

I'd appreciate if you could have another look and let me know if you find anything else that could be improved.

ojeda commented on rust/kernel/sync/mod.rs:
```diff
@@ -0,0 +1,59 @@
+//! This module contains the kernel APIs related to synchronisation that have been ported or
+//! wrapped for usage by Rust code in the kernel and is shared by all of them.
+//!
+//! # Example
```
I'd move the example into `mutex.rs`, i.e. so that we have a different one for condvar, spinlock, etc.

Not sure if it is best to also move each of them to `mutex_init!`, rather than have them in the module doc, hmm... I will think about it, but we can move it later.

wedsonaf replied:
If we move it to `mutex.rs`, would it be visible in the rendered docs? (Given that is gets 'exported' via `pub use`.)

kloenk replied:
There is an attribute for pub use, to render the doc in place, so yes we can rerender it as it would be in this file.

ojeda commented on rust/kernel/sync/mod.rs:
````diff
@@ -0,0 +1,59 @@
+//! # Example
+//!
+//!```
+//! fn test() {
````
No need for wrapping into a `fn test() {` in examples, i.e. the code can go directly in "file" scope.

ojeda commented on rust/kernel/sync/mutex.rs:
```diff
@@ -0,0 +1,101 @@
+    data: UnsafeCell<T>,
+}
+
+// SAFETY: `Mutex` can be transferred across thread boundaries iff the data it protects can.
```
Just in case: the "iff" is "if and only if" or a typo?

wedsonaf replied:
"if and only if"

ojeda reviewed (commented):
Looks very good. None of the comments is really a blocker. I will merge it tomorrow to give time to others to take a second look and then we move into the other two.

wedsonaf reviewed (commented):
> Looks very good. None of the comments is really a blocker. I will merge it tomorrow to give time to others to take a second look and then we move into the other two.

Sounds good, thank you!
